crypto/threads_win.c

```diff
@@ -675,7 +729,8 @@ int CRYPTO_atomic_load_int(int *val, int *ret, CRYPTO_RWLOCK *lock)
 
     return 1;
 #else
-    /* On Windows, LONG is always the same size as int. */
+    /* On Windows, LONG is always the same size as int, */
+    /* except for the mingw64 target where long int is 64 bit. */
```

Comment:
Hmm... should the cast to LONG be avoided? Or what is the function signature of InterlockedOr() on mingw64??

Reply:
I've never used mingw64 for development, so I did a search and found that the declaration is
``` C
// mingw64
__LONG32 _InterlockedOr(__LONG32 volatile *, __LONG32);

// Win32
LONG InterlockedOr(
  [in, out] LONG volatile *Destination,
  [in]      LONG          Value
);
```
In both cases the parameters are 32 bit. There are variants with 8, 16, 32 and 64 bit parameters.
https://github.com/mirror/mingw-w64/blob/eff726c461e09f35eeaed125a3570fa5f807f02b/mingw-w64-headers/include/psdk_inc/intrin-impl.h#L1594
https://learn.microsoft.com/en-us/windows/win32/api/winnt/nf-winnt-interlockedor

I'm talking from memory here: whether the cast can be removed depends on the compiler. Some see that two pointer types hold the same size integer, and accept it. Others complain just because we have two different names for the same type, and require a cast. The rule is that by making a cast we tell the compiler: we know what we are doing. In some cases this is used to hide a warning. Disrespecting pointer sizes is a good way to corrupt memory.

Here is what I usually do on mismatch:
- integer: I would always cast to what the function expects. This makes it obvious when reading the code, that the function expects a different type
- pointer to integer: I would pass a temporary variable, and copy the data to and from. This is a safe approach that prevents memory corruption.

The tricky part here is that we can't just pass a temporary variable, because we intend to perform an atomic operation on the original parameter.

We can keep the cast to hide any warnings, but add a precompiler assertion like
``` C
#if sizeof(*val) != sizeof(LONG *)
#error using a 64 bit version of this function with a 32 bit argument will corrupt memory
#endif
```
Or we can fallback to the `NO_INTERLOCKEDOR64` case and issue a build warning.

Reply:
> add a precompiler assertion

Note: you can't use `sizeof()` in preprocessor conditions - the preprocessor doesn't know about types in the language. You can use `static_assert()` if it's present, though.

I am generally in favour of casting as long as there is a static assert to ensure that the preconditions for the cast to be sound are met.

Reply:
Please do this then:
```
# ifdef __MINGW64__
    *ret = (int)InterlockedOr((__LONG32 volatile *)val, 0);
# else
    *ret = (int)InterlockedOr((LONG volatile *)val, 0);
# endif
```

Reply:
@t8m Done.
I also fixed the result of `atomic_or_uint_nv`, which was improperly casted to `int *` instead of `int`.

Reply:
Did a quick grep:

winbase.h
```c
LONG __cdecl InterlockedOr (LONG volatile *Destination, LONG Value);
```
winnt.h
```c
typedef __LONG32 LONG;
```

So I guess the function signature are identical.

Reply:
@kwkam is that on mingw64? In that case `long int` is 32 bit there as well or LONG is not the same as `long int`. This is really confusing.

Reply:
@t8m Yes, mingw64.

[_mingw.h](https://github.com/mirror/mingw-w64/blob/eff726c461e09f35eeaed125a3570fa5f807f02b/mingw-w64-headers/crt/_mingw.h.in#L18-L27)
```c
/* Target specific macro replacement for type "long".  In the Windows API,
   the type long is always 32 bit, even if the target is 64 bit (LLP64).
   On 64 bit Cygwin, the type long is 64 bit (LP64).  So, to get the right
   sized definitions and declarations, all usage of type long in the Windows
   headers have to be replaced by the below defined macro __LONG32. */
#ifndef __LP64__	/* 32 bit target, 64 bit Mingw target */
#define __LONG32 long
#else			/* 64 bit Cygwin target */
#define __LONG32 int
#endif
```


So it depends on the platform, and mingw64 `long` is 32 bit.

Reply:
So it is a cygwin where the problem is.

Reply:
But yeah the added comment line is then wrong and should be dropped.